### PR TITLE
planner: make index merge union case aware of order property

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/48359

Problem Summary:

### What changed and how does it work?
As shown in the issue, currently index merge path generation can not feel its parent's property, and this kind of generation contains pruning logic inside ahead, so for this case, since ac and bc index pathes here has the same countAfterAccess as index a and b did, so it will only keep the later one remained.

generation should aware of order property and produce the most suitable index partial pathes combination, which means we should move this generation logic a little bit later when physical enumeration take place, in which parent can ask for what kind of property they want (order by c for this case here)

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
planner: make index merge union case aware of order property
```

### CODE REVIEW

File: pkg/planner/core/indexmerge_path.go [hunk 1]
```diff
@@ -180,6 +180,10 @@ func (ds *DataSource) generateNormalIndexPartialPaths4DNF(dnfItems []expression.
 }
 
 // getIndexMergeOrPath generates all possible IndexMergeOrPaths.
+// For index merge union case，the order property from its partial
+// path can be kept and multi-way merged and output. So we don't
+// generate a concrete index merge path out, but an un-determined
+// alternatives set index merge path instead.
```
Comment (qw4990) [inline pkg/planner/core/indexmerge_path.go:186]: Could you add an example here? For example, for `select ... order by ...`, the `partialAlternativePaths` can be `...`.

File: pkg/planner/core/find_best_task.go [hunk 2]
```diff
@@ -835,6 +836,172 @@ func (ds *DataSource) isMatchProp(path *util.AccessPath, prop *property.Physical
 	return isMatchProp
 }
 
+func (ds *DataSource) matchPropForIndexMergeAlternatives(path *util.AccessPath, prop *property.PhysicalProperty) (*util.AccessPath, bool) {
```
Comment (qw4990) [inline pkg/planner/core/find_best_task.go:839]: Could you add an example for this method?

File: pkg/planner/core/find_best_task.go [hunk 3]
```diff
@@ -835,6 +836,172 @@ func (ds *DataSource) isMatchProp(path *util.AccessPath, prop *property.Physical
...
+				return cmp.Compare(lhsCountAfter, rhsCountAfter)
+			})
+		}
+		allMatchIdxes = append(allMatchIdxes, idxWrapper{matchIdxes, pathIdx})
+	}
+	// sort allMatchIdxes by its element length.
+	// index merge:                index merge:
+	//  path1: {pk, index1}  ==>    path2: {pk}
+	//  path2: {pk}                 path1: {pk, index1}
+	// here for the fixed choice pk of path2, let it be the first one to choose, left choice of index1 to path1.
```
Comment (time-and-fate) [inline pkg/planner/core/find_best_task.go:908]: I'm a bit skeptical about whether we really need to deliberately avoid choosing the same index.

File: pkg/planner/util/path.go [hunk 4]
```diff
@@ -51,6 +51,12 @@ type AccessPath struct {
 	// PartialIndexPaths store all index access paths.
 	// If there are extra filters, store them in TableFilters.
 	PartialIndexPaths []*AccessPath
+	// For every dnf/cnf item, there maybe several matched partial index paths to be determined later in property detecting and cost model.
+	PartialAlternativeIndexPaths [][]*AccessPath
+	ShouldBeKeptCurrentFilter    bool
+	// ShouldBeKeptCurrentFilter indicates that there are some expression inside this dnf that couldn't be pushed down, and we should keep the entire dnf above.
+	ShouldBeKeptCurrentFilterExpression expression.Expression
```
Comment (time-and-fate) [inline pkg/planner/util/path.go:58]: Some suggestions about the comments and naming:
1. Make it clear that an access path with `PartialAlternativeIndexPaths` is a different one, and it can't have `PartialIndexPaths`, and a single index or table path also doesn't use this field.
2. Specify that `ShouldBeKeptCurrentFilter` and `ShouldBeKeptCurrentFilterExpression` are only used with `PartialAlternativeIndexPaths`, which means for the new type of access path.
3. "should be kept" sounds a bit strange to me, and so does the `CurrentFilter`. How about changing the names, like `IndexMergeSourceFilter` and `KeepIndexMergeSourceFilter`?

File: pkg/planner/util/path.go [hunk 5]
```diff
@@ -118,6 +124,11 @@ func (path *AccessPath) IsTablePath() bool {
 	return path.IsIntHandlePath || path.IsCommonHandlePath
 }
 
+// IsIndexMergePath returns true if it's Index Merge Path.
+func (path *AccessPath) IsIndexMergePath() bool {
+	return len(path.PartialIndexPaths) > 0 || len(path.PartialAlternativeIndexPaths) > 0
+}
```
Comment (time-and-fate) [inline pkg/planner/util/path.go:130]: Looks like no places are using this.

**Review by time-and-fate — COMMENTED** [review]
I suggest we replace all `SortFunc` with `SortStableFunc` to make it deterministic.

File: pkg/planner/core/indexmerge_path.go [hunk 6]
```diff
@@ -210,48 +214,55 @@ func (ds *DataSource) generateIndexMergeOrPaths(filters []expression.Expression)
...
-			partialPaths = append(partialPaths, partialPath)
-		}
-		// If all of the partialPaths use the same index, we will not use the indexMerge.
-		singlePath := true
-		for i := len(partialPaths) - 1; i >= 1; i-- {
-			if partialPaths[i].Index != partialPaths[i-1].Index {
-				singlePath = false
-				break
+			// 2.1: trade off on countAfterAccess.
+			theoreticalMinCountAfterAccessPath := ds.buildIndexMergePartialPath(partialAlternativePaths[i])
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_path.go:241]:
```suggestion
			minCountAfterAccessPath := ds.buildIndexMergePartialPath(partialAlternativePaths[i])
```
I think we don't need to specially say it's "theoretical".

File: pkg/planner/util/path.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AilinKid) [reply pkg/planner/util/path.go:58]: impressive, thx

File: pkg/planner/core/indexmerge_path.go [hunk 7]
```diff
@@ -210,48 +214,55 @@ func (ds *DataSource) generateIndexMergeOrPaths(filters []expression.Expression)
...
-			possiblePath := ds.buildIndexMergeOrPath(filters, partialPaths, i, shouldKeepCurrentFilter)
+		// 2.2 get the theoretical whole count after access for index merge.
+		accessDNF := expression.ComposeDNFCondition(ds.SCtx(), accessConds...)
+		sel, _, err := cardinality.Selectivity(ds.SCtx(), ds.tableStats.HistColl, []expression.Expression{accessDNF}, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("something wrong happened, use the default selectivity", zap.Error(err))
+			sel = SelectionFactor
+		}
+
+		if len(partialAlternativePaths) > 1 {
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply pkg/planner/core/indexmerge_path.go:259]: not really, `if len(indexMap) == 1 {` will fail it as they are all about single same index

File: pkg/planner/core/find_best_task.go (hunk 3 above)
Comment (AilinKid) [reply pkg/planner/core/find_best_task.go:908]: i think if we have a>2 or a<1, we can directly derive a $(-\infty, 1) \cup (2, +\infty)$ on a simple index range scan of a.
rather than index merge { $a1:  (-\infty, 1),  a2: (2, +\infty)$ }

old logic, just keep them identical here.

File: pkg/planner/core/testdata/index_merge_suite_in.json [hunk 8]
```diff
@@ -39,5 +39,14 @@
       "select /*+ use_index_merge(t3, ia, ibc, id, ie) */ * from t3 where a > 10 and b = 20 and c < 35 and d < 45 and e = 100",
       "select /*+ use_index_merge(t4, ia, ibc, id, ie) */ * from t4 where a > 10 and b = 20 and c < 35 and d in (1,3,8,9) and e = 100"
     ]
+  },
+  {
+    "name": "TestIndexMergeWithOrderProperty",
+    "cases": [
```
Comment (qw4990) [inline pkg/planner/core/testdata/index_merge_suite_in.json:45]: Could you add more cases:
```
-- duplicated column in DNF
where a=1 or a=2 order by c
where a=1 or a=2 or b=3 order by c
-- more than 2 columns
where a=1 or b=1 or c=1 order by d
where a=1 or b=1 or c=1 or d=1 order by e -- idx(a, e), (b, e), ...
-- composite predicates
where (a=1 and b=1) or c=1 order by d  -- idx(a, b, d), idx(c, d)
where (a=1 and b=1) or (c=1 and d=1) order by e
```

**Review by qw4990 — APPROVED** [review]

File: pkg/planner/core/testdata/index_merge_suite_in.json (hunk 8 above)
Comment (AilinKid) [reply pkg/planner/core/testdata/index_merge_suite_in.json:45]: great test

File: pkg/planner/core/find_best_task.go (hunk 3 above)
Comment (time-and-fate) [reply pkg/planner/core/find_best_task.go:971]: That makes sense. However, if so, do we really need this sorting logic? I can't come up with any real cases that need this logic.

File: pkg/planner/core/indexmerge_path.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (time-and-fate) [reply pkg/planner/core/indexmerge_path.go:288]: I mean we write `if len(partialAlternativePaths) == 1 { continue}` there.

File: tests/integrationtest/r/planner/core/issuetest/planner_issue.result [hunk 9]
```diff
@@ -113,10 +113,13 @@ explain select  /*+ use_index_merge( tbl_39) */ col_239  from tbl_39 where not(
...
+    └─IndexMerge_29	382.00	root		type: union
+      ├─Selection_25(Build)	0.05	cop[tikv]		not(in(planner__core__issuetest__planner_issue.tbl_39.col_239, 2004, 2010, 2010))
+      │ └─IndexRangeScan_24	0.14	cop[tikv]	table:tbl_39, index:PRIMARY(col_239)	range:[1994,1994], keep order:true, stats:pseudo
+      ├─Selection_27(Build)	458.26	cop[tikv]		or(lt(cast(planner__core__issuetest__planner_issue.tbl_39.col_239, double UNSIGNED BINARY), 2026), gt(cast(planner__core__issuetest__planner_issue.tbl_39.col_239, double UNSIGNED BINARY), 2011))
+      │ └─IndexRangeScan_26	477.36	cop[tikv]	table:tbl_39, index:idx_223(col_239)	range:(1996,+inf], keep order:true, stats:pseudo
+      └─TableRowIDScan_28(Probe)	382.00	cop[tikv]	table:tbl_39	keep order:false, stats:pseudo
```
Comment (time-and-fate) [inline tests/integrationtest/r/planner/core/issuetest/planner_issue.result:122]: Looks like we are choosing a worse execution plan. I guess this should be a cost model issue?

File: pkg/planner/core/casetest/dag/testdata/plan_suite_out.json [hunk 10]
```diff
@@ -16,7 +16,7 @@
       },
       {
         "SQL": "select * from t where (t.c > 0 and t.c < 2) or (t.c > 4 and t.c < 6) or (t.c > 8 and t.c < 10) or (t.c > 12 and t.c < 14) or (t.c > 16 and t.c < 18)",
-        "Best": "TableReader(Table(t)->Sel([or(or(and(gt(test.t.c, 0), lt(test.t.c, 2)), and(gt(test.t.c, 4), lt(test.t.c, 6))), or(and(gt(test.t.c, 8), lt(test.t.c, 10)), or(and(gt(test.t.c, 12), lt(test.t.c, 14)), and(gt(test.t.c, 16), lt(test.t.c, 18)))))]))"
+        "Best": "IndexLookUp(Index(t.c_d_e)[[1,1] [5,5] [9,9] [13,13] [17,17]], Table(t))"
```
Comment (time-and-fate) [inline pkg/planner/core/casetest/dag/testdata/plan_suite_out.json:19]: Why is this changed? The old and the new execution plans both don't contain `IndexMerge`.

File: pkg/planner/core/indexmerge_intersection_test.go [hunk 11]
```diff
@@ -43,6 +43,40 @@ func TestPlanCacheForIntersectionIndexMerge(t *testing.T) {
...
+func TestIndexMergeWithOrderProperty(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int, c int, d int, e int, key a(a), key b(b), key c(c), key ac(a, c), key bc(b, c), key ae(a, e), key be(b, e)," +
+		" key abd(a, b, d), key cd(c, d))")
+	tk.MustExec("create table t2 (a int, b int, c int, key a(a), key b(b), key ac(a, c))")
+
+	tk.MustExec("")
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_intersection_test.go:55]: ?

File: pkg/planner/core/indexmerge_intersection_test.go [hunk 12]
```diff
@@ -43,6 +43,40 @@ func TestPlanCacheForIntersectionIndexMerge(t *testing.T) {
...
+
+	var (
+		input  []string
+		output []struct {
+			SQL    string
+			Plan   []string
+			Result []string
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_intersection_test.go:62]: We didn't insert any data. So I think we can remove the `Result` field to make the test result cleaner.

File: pkg/planner/core/find_best_task.go (hunk 3 above)
Comment (AilinKid) [reply pkg/planner/core/find_best_task.go:975]: sorting according to valid candidate length here can avoid the miss-pick across paths.
 
in the deriveStats, we already checked the simple index rule before, when generating a concrete path here, we should successfully generate it out, rather than “can't find a proper physical plan” caused by miss-pick here, saying 	
```
// index merge:                index merge:
//  path1: {pk, index1}  ==>    path2: {pk}
//  path2: {pk}                 path1: {pk, index1}
```
this index merge was validated before, when generating concrete paths, we should choose a path for path2 first, otherwise when picking pk for path1 is done, path2 has no choices then. (error)

File: pkg/planner/core/indexmerge_test.go [hunk 13]
```diff
@@ -39,11 +39,19 @@ func getIndexMergePathDigest(paths []*util.AccessPath, startIndex int) string {
 		}
 		path := paths[i]
 		idxMergeDisgest += "{Idxs:["
-		for j := 0; j < len(path.PartialIndexPaths); j++ {
+		for j := 0; j < len(path.PartialAlternativeIndexPaths); j++ {
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply pkg/planner/core/indexmerge_test.go:42]: this functionality is used in test: TestIndexMergePathGeneration, which has made some assertions with the logical plan after deriveStats, where PartialIndexPaths will be nil according to the design in this pull request

File: tests/integrationtest/r/planner/core/issuetest/planner_issue.result (hunk 9 above)
Comment (AilinKid) [reply tests/integrationtest/r/planner/core/issuetest/planner_issue.result:122]: maybe this just show respect to the hint /*+ use_index_merge( tbl_39) */

File: pkg/planner/core/indexmerge_path.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (AilinKid) [reply pkg/planner/core/indexmerge_path.go:298]: make sense

File: pkg/planner/core/find_best_task.go [hunk 14]
```diff
@@ -835,6 +836,235 @@ func (ds *DataSource) isMatchProp(path *util.AccessPath, prop *property.Physical
...
+				found = true
+				break
+			}
+		}
+		if !found {
+			// just pick the same name index (just using the first one is ok), in case that there may be some other
+			// picked distinctive index path for other partial paths latter.
+			determinedIndexPartialPaths = append(determinedIndexPartialPaths, alternatives[matchIdxes.matchIdx[0]].Clone())
+			// uedIndexMap[oneItemAlternatives[oneIdx].Index.ID] = struct{}{} must already be colored.
+		}
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply pkg/planner/core/find_best_task.go:1008]: //  path1: {pk, index1}  ==>    path1: {index1}
//  path2: {pk}              ==>    path2: {pk}  
//  path3: {pk}              ==>    path3: {pk}  

this case is allowable, when picking path for path3, when pk is already colored, we can also pick pk as well.

File: pkg/planner/core/casetest/dag/testdata/plan_suite_out.json (hunk 10 above)
Comment (AilinKid) [reply pkg/planner/core/casetest/dag/testdata/plan_suite_out.json:19]: i'm quite strange about this, after debug inside i find some difference in the estimation of selectivity
image
image
and finally leading the ds.statsCount difference about 1250 vs 50

File: pkg/planner/util/path.go [hunk 15]
```diff
@@ -51,6 +51,19 @@ type AccessPath struct {
...
 	PartialIndexPaths []*AccessPath
+
+	// ************************************************** special field below *********************************************************
+	// For every dnf/cnf item, there maybe several matched partial index paths to be determined later in property detecting and cost model.
+	// when PartialAlternativeIndexPaths is not empty, it means a special state for index merge path, and it can't have PartialIndexPaths
+	// at same time. Normal single index or table path also doesn't use this field.
+	PartialAlternativeIndexPaths [][]*AccessPath
+	// ShouldBeKeptCurrentFilter and ShouldBeKeptCurrentFilterExpression are only used with PartialAlternativeIndexPaths, which means for
+	// the new state/type of access path. (undetermined index merge path)
+	KeepIndexMergeORSourceFilter bool
```
Comment (time-and-fate) [inline pkg/planner/util/path.go:62]: Please update the comments to be consistent with the new variable name.

File: pkg/planner/core/find_best_task.go (hunk 3 above)
Comment (time-and-fate) [reply pkg/planner/core/find_best_task.go:975]: I mean if there are real examples like SQLs?
In such a case, I think, can we just choose a single-index plan on `pk`, or just allow using two partial paths using the same index?

File: tests/integrationtest/r/planner/core/issuetest/planner_issue.result (hunk 9 above)
Comment (time-and-fate) [reply tests/integrationtest/r/planner/core/issuetest/planner_issue.result:122]: Yes. I think you are right.

**Review by time-and-fate — COMMENTED** [review]
Others LGTM.

File: pkg/planner/core/find_best_task.go (hunk 3 above)
Comment (AilinKid) [reply pkg/planner/core/find_best_task.go:975]: > I mean if there are real examples like SQLs? 
the modification here comes from an old unit test case failure, so I changed the logic like this.

the main reason is that: the user specified an index merge hint, then we pruned other paths in deriveStats there, the only left path is validated for index merge choosing (not all a simple index, otherwise, index merge hint is inapplicable right?), so the concrete path should successfully be chosen by a kind of order that we want before.

**Review by time-and-fate — APPROVED** [review]
